### PR TITLE
date: Implement setting the date on Unix & Windows

### LINKED ISSUE #1754: date: implement the set date
https://github.com/uutils/coreutils/blob/bfc9e13f113f11be2762fe0f72d74628db88fd36/src/uu/date/src/date.rs#L234  isn't implemented.

It should be cross platform but starting with one platform first is fine.

Tests are in
https://github.com/uutils/coreutils/blob/8337179454d0c952746a980ee37a32340f510a46/tests/by-util/test_date.rs

### PR DESCRIPTION
**Pull request by marcosatti**
Fixes #1754

Implements basic functionality for setting the date under both Unix & Windows systems.

Unresolved issues:
- `parse_date` only accepts limited formats for the dates input - separate issue needed?
- How are leap seconds handled on Windows? [Not immediately clear](https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/support-for-leap-second), seems like they kind of ignore it which is what I've put in.

### CODE REVIEW

**Comment by sylvestre** [comment]
It would be great to add tests too.
I guess it requires root but we could:
* test that the parsing of the input works
* test that the parsing of the input fails with the right error
* setting the date as not root fails the expected way :)

thanks

File: src/uu/date/src/date.rs [hunk 1]
```diff
@@ -12,13 +12,20 @@
...
 
+use chrono::{DateTime, FixedOffset, Local, Offset, Utc};
+#[cfg(windows)]
+use chrono::{Datelike, Timelike};
 use clap::{App, Arg};
-
-use chrono::offset::Utc;
-use chrono::{DateTime, FixedOffset, Local, Offset};
+#[cfg(unix)]
+use libc::{clock_settime, timespec, CLOCK_REALTIME};
```
[the remark this answers is not in the document]
Comment (marcosatti) [reply src/uu/date/src/date.rs:20]: Appears to be no way to set the clock on mac systems, so I'll go with disabling the feature like you said.

**Comment by sylvestre** [comment]
Has some builds failures!

**Comment by marcosatti** [comment]
> 
> 
> Has some builds failures!

Yeah not quite completed yet... 

The 32-bit platforms are still using `long` as `time_t`'s, so I'll jut add in a `try_into()` call - they should be updated by 2038 I would hope. The nanoseconds can fit into a i32 no matter what, so a static cast should be fine there.

**Comment by marcosatti** [comment]
The gnu date always outputs the current date (regardless if there was a permissions error) when setting the date - do you want to create a new issue for this or should it be done in this one?

**Comment by marcosatti** [comment]
Damn, try_into() was added in a later Rust version (v1.34?)... I'll just use a normal cast.

**Comment by sylvestre** [comment]
please run
```
cargo +1.33.0 update
```

File: src/uu/date/src/date.rs [hunk 2]
```diff
@@ -294,6 +303,23 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
     0
 }
 
+#[cfg(any(windows, all(unix, not(target_os = "macos"))))]
+/// Apply platform specific matcher arguments to the app given.
+fn make_platform_app<'a, 'b>(app: App<'a, 'b>) -> App<'a, 'b> {
```
Comment (sylvestre) [inline src/uu/date/src/date.rs:308]: I wonder if this isn't overkill ?
I think the option should also exist on mac but in the help section, add "not available on mac yet"
and this function should be added back in the main app declaration

File: src/uu/date/src/date.rs [hunk 3]
```diff
@@ -314,3 +340,73 @@ fn make_format_string(settings: &Settings) -> &str {
...
+/// If it fails, return a tuple of the `String` along with its `ParseError`.
+fn parse_date<S: AsRef<str> + Clone>(
+    s: S,
+) -> Result<DateTime<FixedOffset>, (String, chrono::format::ParseError)> {
+    // TODO: The GNU date command can parse a wide variety of inputs.
+    s.as_ref().parse().map_err(|e| (s.as_ref().into(), e))
+}
+
+/// Displays the errno string to stderr and returns the error code.
+fn get_errno() -> i32 {
```
Comment (sylvestre) [inline src/uu/date/src/date.rs:354]: same, should not be declared for mac

File: src/uu/date/src/date.rs [hunk 4]
```diff
@@ -314,3 +340,73 @@ fn make_format_string(settings: &Settings) -> &str {
...
+
+    if result != 0 {
+        get_errno()
+    } else {
+        0
+    }
+}
+
+#[cfg(windows)]
+/// System call to set date (Windows).
```
Comment (sylvestre) [inline src/uu/date/src/date.rs:387]: thanks for the comments :)

File: tests/by-util/test_date.rs [hunk 5]
```diff
@@ -131,3 +133,34 @@ fn test_date_format_full_day() {
     let re = Regex::new(r"\S+ \d{4}-\d{2}-\d{2}").unwrap();
     assert!(re.is_match(&result.stdout.trim()));
 }
+
+#[test]
+#[cfg(all(unix, not(target_os = "macos")))]
+fn test_date_set_valid() {
```
Comment (sylvestre) [inline tests/by-util/test_date.rs:139]: please rustfmt

**Review by sylvestre — CHANGES_REQUESTED** [review]
only minor requests!
thanks

**Comment by sylvestre** [comment]
@marcosatti ping?

**Comment by marcosatti** [comment]
@sylvestre sorry been busy lately - will look at it later on in the week.

File: src/uu/date/src/date.rs [hunk 6]
```diff
@@ -314,3 +330,76 @@ fn make_format_string(settings: &Settings) -> &str {
         Format::Default => "%c",
     }
 }
+
+/// Parse a `String` into a `DateTime`.
+/// If it fails, return a tuple of the `String` along with its `ParseError`.
+fn parse_date<S: AsRef<str> + Clone>(
+    s: S,
+) -> Result<DateTime<FixedOffset>, (String, chrono::format::ParseError)> {
+    // TODO: The GNU date command can parse a wide variety of inputs.
```
Comment (sylvestre) [inline src/uu/date/src/date.rs:339]: In the tests, could you please add a case that GNU manages but we don't?

**Comment by sylvestre** [comment]
Looks great :)
I will merge yet now, please open a PR for the last request!

**Comment by marcosatti** [comment]
Created PR #1846 for more tests.

**Pull request merged**